src/core/net/ip6.cpp

```diff
@@ -603,11 +614,303 @@ otError Ip6::HandleOptions(Message &aMessage, Header &aHeader, bool &aForward)
+        SuccessOrExit(error = message->SetOffset(0));
+        message->SetDatagramTag(fragmentHeader.GetIdentification());
+
+        // copying the non-fragmentable header to the fragmentation buffer
+        assertValue = aMessage.CopyTo(0, 0, aMessage.GetOffset(), *message);
+        assert(assertValue == aMessage.GetOffset());
+
+        if (!mTimer.IsRunning())
+        {
+            mTimer.Start(kStateUpdatePeriod);
+        }
+
+        mReassemblyList.Enqueue(*message);
+
+        otLogDebgIp6("start reassembly.");
+    }
+
+    otLogInfoIp6("Fragment with id %d received > %d bytes, offset %d", message->GetDatagramTag(), payloadFragment,
+                 offset);
+
+    if (offset + payloadFragment + aMessage.GetOffset() > kMaxAssembledDatagramLength)
+    {
+        otLogWarnIp6("Package too large for fragment buffer");
+        ExitNow(error = OT_ERROR_NO_BUFS);
+    }
+
+    // increase message buffer if necessary
+    if (message->GetLength() < offset + payloadFragment + aMessage.GetOffset())
+    {
+        SuccessOrExit(error = message->SetLength(offset + payloadFragment + aMessage.GetOffset()));
+    }
+
+    // copy the fragment payload into the message buffer
+    assertValue = aMessage.CopyTo(aMessage.GetOffset() + sizeof(fragmentHeader), aMessage.GetOffset() + offset,
+                                  payloadFragment, *message);
+    assert(assertValue == static_cast<int>(payloadFragment));
+
+    // check if it is the last frame
+    if (fragmentHeader.IsMoreFlagSet() == false)
+    {
+        // use the offset value for the whole ip message length
+        SuccessOrExit(error = message->SetOffset(offset + payloadFragment + aMessage.GetOffset()));
+    }
+
+    if (message->GetOffset() >= message->GetLength())
+    {
+        offset = 0;
+
+        // creates the header for the reassembled ipv6 package
+        VerifyOrExit(aMessage.Read(0, sizeof(header), &header) == sizeof(header), error = OT_ERROR_PARSE);
+        header.SetPayloadLength(message->GetLength() - sizeof(header));
+        header.SetNextHeader(static_cast<IpProto>(fragmentHeader.GetNextHeader()));
+        assertValue = message->Write(0, sizeof(header), &header);
+        assert(assertValue == sizeof(header));
+
+        otLogDebgIp6("Reassembly complete.");
+
+        mReassemblyList.Dequeue(*message);
+
+        error = HandleDatagram(*message, aNetif, aMessageInfo.mLinkInfo, aFromNcpHost);
+    }
+
+exit:
+    if (error != OT_ERROR_DROP && error != OT_ERROR_NONE && isFragmented)
+    {
+        if (message != NULL)
+        {
+            mReassemblyList.Dequeue(*message);
+            message->Free();
+        }
+        otLogWarnIp6("Reassembly failed: %s", otThreadErrorToString(error));
+    }
+    if (isFragmented)
+    {
+        // drop all fragments, the payload is stored in the fragment buffer
+        error = OT_ERROR_DROP;
+    }
+
+    return error;
+}
+
+void Ip6::CleanupFragmentationBuffer(void)
+{
+    Message *message;
+    Message *next;
```

Comment:
nit: prefer declaring in smallest enclosing scope

Reply:
done